### PR TITLE
Ask SChannel for stapled OCSP when the client will do revocation

### PR DESCRIPTION
**Pull request by bartonjs**
I haven't been able to diagnose why Windows in the Server role doesn't seem to want to participate in OCSP Stapling (at least during tests), but with this change the clients will get the benefits of stapled OCSP when the server participates.

The change essentially boils down to:
* Set SCH_CRED_REVOCATION_CHECK_END_CERT if we're going to ask for revocation (so SChannel sends the status_request:OCSP extension in the handshake)
* Test that we see certificate property 70 on Windows when we expect to.

Everything else is plumbing to make that happen.

Contributes to #33377.

### CODE REVIEW

**Comment by wfurt** [comment]
I'm OOF @bartonjs. I should be able to look at it early next week.

File: src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs [hunk 1]
```diff
@@ -134,6 +135,62 @@ public Task ConnectWithRevocation_ServerCertWithoutContext_NoStapledOcsp()
             return ConnectWithRevocation_WithCallback_Core(X509RevocationMode.Offline, offlineContext: null);
         }
 
+#if WINDOWS
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs:138]: What does this use that won't compile in a non-Windows build?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 2]
```diff
@@ -196,17 +202,29 @@ public static unsafe SafeFreeCredentials AcquireCredentialsHandleSchannelCred(Ss
...
                     Interop.SspiCli.SCHANNEL_CRED.Flags.SCH_CRED_NO_DEFAULT_CREDS |
                     Interop.SspiCli.SCHANNEL_CRED.Flags.SCH_SEND_AUX_RECORD;
+
+                // Request OCSP Stapling from the server
+                if (authOptions.CertificateRevocationCheckMode != X509RevocationMode.NoCheck)
+                {
+                    flags |=
+                        Interop.SspiCli.SCHANNEL_CRED.Flags.SCH_CRED_REVOCATION_CHECK_END_CERT |
+                        Interop.SspiCli.SCHANNEL_CRED.Flags.SCH_CRED_IGNORE_NO_REVOCATION_CHECK |
+                        Interop.SspiCli.SCHANNEL_CRED.Flags.SCH_CRED_IGNORE_REVOCATION_OFFLINE;
```
Comment (wfurt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:212]: If the mode is X509RevocationMode.Online should we ignore failures? While that may be convenient I'm wondering if that meets security expectations.

File: src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs [hunk 3]
```diff
@@ -134,6 +135,62 @@ public Task ConnectWithRevocation_ServerCertWithoutContext_NoStapledOcsp()
...
+        [Theory]
```
Comment (wfurt) [inline src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs:139]: This needs to be `ConditionalTheory` for `SkipTestException` to work.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs (hunk 2 above)
Comment (bartonjs) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:212]: We're ignoring any failures out of SChannel anyways, and we rebuild the chain using the data they provide us.

**Review by stephentoub — APPROVED** [review]

**Review by wfurt — APPROVED** [review]
LGTM

**Pull request merged**